morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs

```diff
@@ -315,71 +491,768 @@ instance
+
+instance MPretty MorphoTransactionError where
+  mpretty MorphoCandidateBeforeCheckpoint = "Vote occurs before latest checkpoint"
+  mpretty MorphoAlreadyCheckpointed = "Vote is already a checkpoint"
+  mpretty MorphoWrongDistance = "Vote is at wrong distance"
+  mpretty MorphoInvalidSignature = "Vote has invalid signature"
+  mpretty MorphoDuplicateVote = "Duplicated vote"
+  mpretty MorphoUnknownPublicKey = "Vote has an unknown public key"
+
+instance MPretty (HeaderError (MorphoBlock h c)) where
+  mpretty (HeaderProtocolError err) = mpretty err
+  mpretty (HeaderEnvelopeError err) = mpretty err
+
+instance MPretty BftValidationErr where
+  mpretty (BftInvalidSignature signature) =
+    "BFT invalid signature:" <+> pretty signature
+
+instance MPretty (HeaderEnvelopeError (MorphoBlock h c)) where
+  mpretty (UnexpectedBlockNo expectedBlockNo actualBlockNo) =
+    "Invalid block number. Expected"
+      <+> mpretty expectedBlockNo
+      <+> "but got"
+      <+> mpretty actualBlockNo
+  mpretty (UnexpectedSlotNo expectedSlotNo actualSlotNo) =
+    "Invalid slot number. Expected minimum slot number of"
+      <+> mpretty expectedSlotNo
+      <+> "but got"
+      <+> mpretty actualSlotNo
+  mpretty (UnexpectedPrevHash oldTip prevHash) =
+    "Invalid hash (in the reference to the previous block). The current tip is"
+      <+> mpretty oldTip
+      <+> "and the previous hash of the new block is"
+      <+> mpretty prevHash
+
+instance MPretty DiskSnapshot where
   mpretty = viaShow
 
-instance MPretty NtC.HandshakeTr where
+instance MPretty ImmutableDB.ChunkNo where
+  mpretty (ImmutableDB.ChunkNo n) = pretty n
+
+instance MPretty VolatileDB.BlockOffset where
+  mpretty (VolatileDB.BlockOffset o) = pretty o
+
+instance MPretty (HeaderHash a) => MPretty (ImmutableDB.Tip a) where
+  mpretty (ImmutableDB.Tip slot isEBB blockNo hash) =
+    "ImmutableDBTip(slot:"
+      <+> mpretty slot
+      <> ", isEpochBoundaryBlock:"
+      <+> mpretty isEBB
+      <> ", hash:"
+      <+> mpretty hash
+      <> ", blockNo:"
+      <+> mpretty blockNo
+      <> ")"
+
+instance MPretty IsEBB where
+  mpretty IsEBB = "true"
+  mpretty IsNotEBB = "false"
+
+instance MPretty FsPath where
   mpretty = viaShow
 
+instance MPretty (TraceEventMempool (MorphoBlock h c)) where
+  mpretty
+    ( TraceMempoolAddedTx
+        tx
+        (MempoolSize numTxsBefore numBytesBefore)
+        (MempoolSize numTxsAfter numBytesAfter)
+      ) =
+      "Transaction"
+        <+> mpretty tx
+        <+> "added to mempool. Number of transactions in mempool goes from"
+        <+> pretty numTxsBefore
+        <+> "to"
+        <+> pretty numTxsAfter
+        <> ". Bytes of mempool goes from"
+        <+> pretty numBytesBefore
+        <+> "to"
+        <+> pretty numBytesAfter
+  mpretty (TraceMempoolRejectedTx tx err (MempoolSize numTxs numBytes)) =
+    "Transaction"
+      <+> mpretty tx
+      <+> "rejected because:"
+      <+> mpretty err <> ". Mempool has"
```

Comment:
```suggestion
      <+> mpretty (map snd err) <> ". Mempool has"
```

Type of `err` is `(Vote, MorphoTransactionError)`, so it carries the vote with it, but in this message we already have the vote in the `tx`, so we can only look at the second value here.